### PR TITLE
[Proposal] Admin UI with SSR & build targets (powered by Next.js)

### PR DESCRIPTION
**Pull request by jesstelford**
This is very WIP, but it works!

_Depends on #871_

admin-ui-next
_(this gif is **not** sped up)_

## Changes

- **This is just a proposal.** If we don't like it, we throw it out. If it works and doesn't introduce any major issues, I'd love to see it merged.
- _**The way a Keystone app developer starts/runs their instances has not changed**_.
- Adds the ability to generate a production UI bundle + server.
- Improves [TTI](https://developers.google.com/web/tools/lighthouse/audits/time-to-interactive).
- Dramatically improves [Time To First Contentful Paint](https://developers.google.com/web/tools/lighthouse/audits/first-contentful-paint).
- Adds a `.isSortable()` & `.isEditable()` API to fields which affects their usage within the Admin UI:
  - `.isSortable()`: If truthy, will enable the clickable titles when in table view for sorting rows _(default: `true`)_. Allows opting-out of sorting when not appropriate, for example; the `Password` field, or an `Avatar` field.
  - `.isEditable()`: If truthy will show the field when editing / creating items in the AdminUI _(default: `true`)_. Allows opting out of editing, for example the `Pseudolabel` field, or a custom field type in the future which is display-only.
- Changes some internals of the AdminUI _(which have no impact on developers of Keystone apps)_:
  - Implements lazy loading of Fields, Cells, and Filters so only the ones that are actually rendered on the page are loaded (eg; the "Date" `Filter.js` wont be loaded unless you click to expand a date filter)
  - Uses `next` to power the Admin UI.
    - It's actually 2 separate next.js instances; 1 for the login and 1 for the admin to 100% prevent any data leakage for non-authenticated users (because all the list info is embedded into the Admin UI's client bundles).
  - Removes all the special casing of `_label_`, replacing it with an actual `Field` called `Pseudolabel` so it's just like everything else.
  - Makes all the `@arch-ui` components SSR compatible.
  - Swaps out usage of `React.createPortal()` for [`@jesstelford/react-portal-universal`](https://github.com/zeit/next.js/tree/canary/examples/with-portals-ssr)
  - Removed `react-router` dependency from the `@arch-ui` packages so they're easier to use universally.

## Notes

- I haven't fixed the tests (or lint or formatting or types)
- `Content` type doesn't render correctly (need to add lazy loading of the _blocks_)
- The `Style Guide` page has not been ported to use next (it currently wont render if accessed)
- There are a lot of white-space changes
- We need to tweak the loading states so they're a little less jarring while lazy loading some things (especially on a slow network)
- Not all the projects are configured correctly to run. For now, the easiest way to test is `cd test-projects/basic && yarn dev`
- The Custom types (`Stars`) demo has been disabled because the code is un-transpiled. We need to do something similar to https://github.com/keystonejs/keystone-5/pull/843 to solve it.

### CODE REVIEW

File: packages/fields/types/CalendarDay/index.js [hunk 1]
```diff
@@ -3,16 +3,15 @@ const {
...
   implementation: CalendarDay,
   views: {
-    Controller: path.join(__dirname, './Controller'),
-    Field: path.join(__dirname, './views/Field'),
-    Filter: path.join(__dirname, './views/Filter'),
-    Cell: path.join(__dirname, './views/Cell'),
+    Controller: '@keystone-alpha/fields/types/CalendarDay/Controller',
+    Field: '@keystone-alpha/fields/types/CalendarDay/views/Field',
+    Filter: '@keystone-alpha/fields/types/CalendarDay/views/Filter',
+    Cell: '@keystone-alpha/fields/types/CalendarDay/views/Cell',
```
[the remark this answers is not in the document]
Comment (mitchellhamilton) [reply packages/fields/types/CalendarDay/index.js]: This explanation helps massively, thanks! ❤️  I have a feeling there's a better way to solve the problems but not gonna look into it until we decide that next is what we're doing.

Also, the reason why I'm concerned about this is that making the views not absolute/not resolve from the package is that someone could have multiple different versions of @keystone-alpha/fields and the admin ui could import the views from the wrong package.

**Comment by mitchellhamilton** [comment]
I want to try to state my big concerns with using Next and SSR in (hopefully) clear terms. Obligatory, please tell me if something I've said is completely wrong or you think I've incorrectly made an assumption or anything else.  Also, it'd be great to get some of the unrelated parts of this PR out like the Label field stuff because all that stuff seems great and is unrelated to everything else.
# performance

this is the main reason you want to do SSR/use next, right? (ignoring static building since #822 shows we can easily do it without next)

I ran lighthouse(default settings, mobile, 3g, 4x cpu slowdown) on #749 with NODE_ENV=production and the results are quite good, especially given how little effort and complexity it took so I want to ask, do you think that the performance improvements that SSR could have are worth the added complexity by Next and SSR?

Note that I also added gzipping which helped a bunch and I think that's a great example of the kind of performance things that I personally think we should be focusing on, low effort, high reward.

lighthouse results

I wanted to run lighthouse on this PR to compare but I couldn't figure out how to start it in production. (running build then start on the basic test project didn't work)

# complexity of using next and SSR in general

> >  the mid-to-long term vision for the Admin UI is that it is a framework for people to build on, not just a product we're in control of.
>
> From a pragmatic point of view; this is now how it's been built today. Far from it, in fact. To move to this kind of model would require significant rewriting of the AdminUI. If we merge this PR, then get to that point, we can always bail out of using Next. At the time we're doing that huge refactor it would be a drop in the bucket in terms of effort. 

Adding things like extra pages from user supplied react component's and letting people replace certain UI elements could work just like we have the field types right now(as you mention). And while we could still do a lot(probably all?) of this with Next, I believe it would make it more complex for people to add custom things because they'd have to think about SSR and as @JedWatson mention's, things like css-in-js libraries, data fetching libraries, etc. require digging into next. (As a concrete, likely common example, you could have a custom page that combines data from keystone with data from an external service)

The solution to this could be use the `noSSR: true` option as you mention but I'd like to ask the question then, what are we getting from next if we're going to disable ssr for things which (IMO) should be common and well supported use cases?

inb4: next's well maintained webpack config, I don't think this is really a benefit so far, I can only see problems so far (https://github.com/keystonejs/keystone-5/pull/867#discussion_r268013572) and no concrete benefits for users or us as people who make keystone since our webpack config right now is very simple and can make a pretty fast app. Or am I missing a benefit from using next's webpack config?

**Comment by jesstelford** [comment]
To avoid spinning wheels on something which looks like it wont land, I'll close this PR now, and we can focus on other things that are more important; like trying to make a website which somehow supports every conceivable css-in-js library and custom pages at the same time.

**Comment by jesstelford** [comment]
# Performance

Here are some comparisons between `master` @ 7f6ccc91 (incl. suspense), and this branch.

All were run with `NODE_ENV=production`, mobile with 4x slowdown.

suspense-dashboard
next-dashboard


---

suspense-users
next-users



---

suspense-user
next-user


---

_And here are all the audits themselves: [`lighthouse-audits.zip`](https://github.com/keystonejs/keystone-5/files/3005964/lighthouse-audits.zip)_

Points to note:

- The TTI is slower by a second in next.
- The first meaningful and contentful paints are over a second faster in next. This has a huge impact on perceived performance. It gives the user that extra time to be scanning the page for the information they want / moving their mouse towards the target they want to click, negating any effects of a slightly slower TTI.


# Complexity of using next and SSR in general

> because they'd have to think about SSR

That's a good thing. It's similar to thinking about accessibility; sure, it's something extra that needs to be considered, but it only results in a better outcome overall.

Also, you can see from this branch that the changes to get things SSR-compatible were minimal (removing a couple `window` checks, and only polyfilling `intersection-observer` on the browser), and this is quite a complex app already.

> things like css-in-js libraries, data fetching libraries, etc. require digging into next.

There's nothing `next`-y about adding a css-in-js lib, or doing data fetching, etc. If you want them to be SSR'd, then, sure you'd have to add that. But we can give them the escape hatch of `noSSR: true`, and they'd be no worse off than what's currently in `master`. But in the majority case; they'd be better off since SSR comes for free.

> I can only see problems so far (#867 (comment)) and no concrete benefits for users or us as people who make keystone since our webpack config right now is very simple and can make a pretty fast app. Or am I missing a benefit from using next's webpack config?

In all the projects I've been involved in; the ones which leant on 3rd parties to handle the complexity always faired better; they are easier to maintain / upgrade, you get new features for free (OSS), and it's easier to onboard folks by pointing them at the docs for the other thing instead of having to write your own (often incomplete) docs.

**Comment by mitchellhamilton** [comment]
> because they'd have to think about SSR
>
>> That's a good thing. It's similar to thinking about accessibility; sure, it's something extra that needs to be considered, but it only results in a better outcome overall.

I don't understand how this is similar to thinking about accessibility, with accessibility it always makes things better but how does having to think about SSR if you're not doing SSR result in a better overall outcome?

> Also, you can see from this branch that the changes to get things SSR-compatible were minimal (removing a couple window checks, and only polyfilling intersection-observer on the browser), and this is quite a complex app already.

The app certainty runs but there are some broken things, many of which can of course be fixed and just require a small amount of effort but the navbar and search/filters/etc. store stuff in local storage and require it to render which means we can't access it during SSR. Right now, this PR just doesn't look at the things in local storage(i assume that was just a temporary fix to get it to generally run?) but if it did look at the things in local storage on the client, the user would see the ui based on the default state from the server and then when all the js loads and mounts, it would change. For example, a user could be about to click on something and then the js mounts and the element they were about to click on could move and then they would click on the wrong element. Do you have any ideas on how this problem could be solved? 

☝️ is my biggest concern, if there's a good solution to that and @JedWatson and others are good with everything, I'd probably be okay with using Next/SSR.

> In all the projects I've been involved in; the ones which leant on 3rd parties to handle the complexity always faired better; they are easier to maintain / upgrade, you get new features for free (OSS), and it's easier to onboard folks by pointing them at the docs for the other thing instead of having to write your own (often incomplete) docs.

I completely 100% agree with that in general but I'm having trouble understanding what complexity we're talking about here, as in, what complexity do we currently deal with ourselves right now that next would solve for us?

**Comment by jesstelford** [comment]
> Do you have any ideas on how [the flash-of-pre-`localstorage`] problem could be solved?

With cookies 👍

**Pull request closed without merging**